diff -r -c krb5-appl-1.0.1/gssftp/ftp/cmds.c krb5-appl-1.0.1_mm/gssftp/ftp/cmds.c
*** krb5-appl-1.0.1/gssftp/ftp/cmds.c	2009-11-05 20:15:06.000000000 +0000
--- krb5-appl-1.0.1_mm/gssftp/ftp/cmds.c	2010-07-03 13:13:48.000000000 +0100
***************
*** 1268,1273 ****
--- 1268,1292 ----
  }
  
  /*
+  * Toggle fwccc mode on server.
+  */
+ /*VARARGS*/
+ void
+ setfwccc()
+ {
+ 
+        if (command ("FWCCC") == COMPLETE) {
+          allowfwccc = !allowfwccc;
+          printf("FWCCC mode %s.\n", onoff(allowfwccc));
+          code = allowfwccc;
+        } else {
+          allowfwccc = 0;
+          printf("We don't support FWCCC mode, sorry.\n");
+          code = -1;
+        }
+ }
+ 
+ /*
   * Toggle metacharacter interpretation
   * on local file names.
   */
diff -r -c krb5-appl-1.0.1/gssftp/ftp/cmdtab.c krb5-appl-1.0.1_mm/gssftp/ftp/cmdtab.c
*** krb5-appl-1.0.1/gssftp/ftp/cmdtab.c	2009-07-08 22:22:07.000000000 +0100
--- krb5-appl-1.0.1_mm/gssftp/ftp/cmdtab.c	2010-07-03 13:14:52.000000000 +0100
***************
*** 117,122 ****
--- 117,123 ----
  char	umaskhelp[] =	"get (set) umask on remote side";
  char	userhelp[] =	"send new user information";
  char	verbosehelp[] =	"toggle verbose mode";
+ char    fwccchelp[] =   "toggle fwccc mode";
  #ifndef NO_PASSIVE_MODE
  char	setpassivehelp[] = "toggle passive transfer mode";
  #endif
***************
*** 203,208 ****
--- 204,210 ----
  	{ "user",	userhelp,	0,	1,	1,	user },
  	{ "umask",	umaskhelp,	0,	1,	1,	do_umask },
  	{ "verbose",	verbosehelp,	0,	0,	0,	setverbose },
+         { "fwccc",      fwccchelp,      0,      0,      0,      setfwccc },
  	{ "?",		helphelp,	0,	0,	1,	help },
  	{ 0 },
  };
diff -r -c krb5-appl-1.0.1/gssftp/ftp/ftp.c krb5-appl-1.0.1_mm/gssftp/ftp/ftp.c
*** krb5-appl-1.0.1/gssftp/ftp/ftp.c	2009-11-20 15:29:22.000000000 +0000
--- krb5-appl-1.0.1_mm/gssftp/ftp/ftp.c	2010-07-03 13:20:37.000000000 +0100
***************
*** 491,496 ****
--- 491,497 ----
  	va_list ap;
  	int r;
  	sig_t oldintr;
+         char* tmpa;
  
  	abrtflag = 0;
  	if (debug) {
***************
*** 517,523 ****
--- 518,534 ----
  again:	if (secure_command(in) == 0)
  		return(0);
  	cpend = 1;
+         if (allowfwccc &&
+            ( !strncmp(fmt, "PORT",4) || !strncmp(fmt, "PASV",4) ) ) {
+                 tmpa = auth_type;
+                 auth_type = NULL;
+         }
  	r = getreply(!strcmp(fmt, "QUIT"));
+         if (allowfwccc &&
+            ( !strncmp(fmt, "PORT",4) || !strncmp(fmt, "PASV",4) ) ) {
+                 auth_type = tmpa;
+                 tmpa=NULL;
+         }
  #ifndef NOENCRYPTION
  	if (r == 533 && clevel == PROT_P) {
  		fprintf(stderr,
***************
*** 1399,1404 ****
--- 1410,1416 ----
  	int on = 1;
  #ifndef NO_PASSIVE_MODE
  	int a1,a2,a3,a4,p1,p2;
+         int sl;
  
  	if (passivemode) {
  		data = socket(AF_INET, SOCK_STREAM, 0);
***************
*** 1409,1419 ****
--- 1421,1436 ----
  		if (options & SO_DEBUG &&
  		    setsockopt(data, SOL_SOCKET, SO_DEBUG, (char *)&on, sizeof (on)) == SOCKET_ERROR)
  			PERROR_SOCKET("ftp: setsockopt (ignored)");
+                 sl = clevel;
+                 if (allowfwccc )
+                         clevel = PROT_C;
  		if (command("PASV") != COMPLETE) {
  			printf("Passive mode refused.  Turning off passive mode.\n");
  			passivemode = 0;
+                         clevel = sl;
  			return initconn();
  		}
+                 clevel = sl;
  
  /*
   * What we've got at this point is a string of comma separated
***************
*** 1484,1493 ****
--- 1501,1514 ----
  		a = (char *)&data_addr.sin_addr;
  		p = (char *)&data_addr.sin_port;
  #define	UC(b)	(((int)b)&0xff)
+                 sl = clevel;
+                 if (allowfwccc )
+                         clevel = PROT_C;
  		result =
  		    command("PORT %d,%d,%d,%d,%d,%d",
  		      UC(a[0]), UC(a[1]), UC(a[2]), UC(a[3]),
  		      UC(p[0]), UC(p[1]));
+                 clevel = sl;
  		if (result == ERROR && sendport == -1) {
  			sendport = 0;
  			tmpno = 1;
***************
*** 1718,1723 ****
--- 1739,1745 ----
  	int prox_type, nfnd;
  	char *volatile cmd2;
  	 fd_set mask;
+         int sl;
  
  	if (strcmp(cmd, "RETR"))
  		cmd2 = "RETR";
***************
*** 1731,1740 ****
--- 1753,1767 ----
  	}
  	if (curtype != prox_type)
  		changetype(prox_type, 1);
+         sl = clevel;
+         if (allowfwccc)
+                 clevel = PROT_C;
  	if (command("PASV") != COMPLETE) {
  		printf("proxy server does not support third party transfers.\n");
+                 clevel = sl;
  		return;
  	}
+         clevel = sl;
  	pswitch(0);
  	if (!connected) {
  		printf("No primary connection\n");
***************
*** 1744,1753 ****
--- 1771,1785 ----
  	}
  	if (curtype != prox_type)
  		changetype(prox_type, 1);
+         sl = clevel;
+         if (allowfwccc)
+                 clevel = PROT_C;
  	if (command("PORT %s", pasv) != COMPLETE) {
  		pswitch(1);
+                 clevel = sl;
  		return;
  	}
+         clevel = sl;
  	if (setjmp(ptabort))
  		goto die;
  	oldintr = signal(SIGINT, abortpt);
diff -r -c krb5-appl-1.0.1/gssftp/ftp/ftp_var.h krb5-appl-1.0.1_mm/gssftp/ftp/ftp_var.h
*** krb5-appl-1.0.1/gssftp/ftp/ftp_var.h	2009-11-05 20:15:06.000000000 +0000
--- krb5-appl-1.0.1_mm/gssftp/ftp/ftp_var.h	2010-07-03 13:22:00.000000000 +0100
***************
*** 78,83 ****
--- 78,84 ----
  extern int	sendport;	/* use PORT cmd for each data connection */
  extern int	verbose;	/* print messages coming back from server */
  extern int	connected;	/* connected to server */
+ extern int      allowfwccc;     /* Allow PORT/PASV/EPRT/EPSV command in clear for GSSAPI*/
  extern int	fromatty;	/* input is from a terminal */
  extern int	interactive;	/* interactively prompt on m* cmds */
  extern int	debug;		/* debugging level */
***************
*** 208,213 ****
--- 209,215 ----
  void settrace (void);
  void sethash (void);
  void setverbose (void);
+ void setfwccc (void);
  void setport (void);
  void setprompt (void);
  void setglob (void);
diff -r -c krb5-appl-1.0.1/gssftp/ftp/main.c krb5-appl-1.0.1_mm/gssftp/ftp/main.c
*** krb5-appl-1.0.1/gssftp/ftp/main.c	2009-11-05 20:15:06.000000000 +0000
--- krb5-appl-1.0.1_mm/gssftp/ftp/main.c	2010-07-03 13:22:43.000000000 +0100
***************
*** 122,127 ****
--- 122,128 ----
  	}
  	doglob = 1;
  	interactive = 1;
+         allowfwccc = 0;
  	autoauth = 1;
  	autologin = 1;
  	forward = 0;
***************
*** 168,173 ****
--- 169,178 ----
  				autoencrypt = 1;
  				break;
  
+                         case 'c':
+                                 allowfwccc = 1;
+                                 break;
+  
  			default:
  			  fprintf(stderr,
  				  "ftp: %c: unknown option\n", *cp);
