diff -rc yafc-1.0/lib/gssapi.c yafc-1.0-mm/lib/gssapi.c
*** yafc-1.0/lib/gssapi.c	2003-04-17 12:52:11.000000000 +0100
--- yafc-1.0-mm/lib/gssapi.c	2003-10-20 18:02:03.000000000 +0100
***************
*** 34,39 ****
--- 34,40 ----
  #include "syshdr.h"
  #include "ftp.h"
  #include "base64.h"
+ #include "gvars.h"
  
  #if defined(HAVE_GSSAPI_H)
  # include <gssapi.h>
***************
*** 241,247 ****
      int context_established = 0;
      char *p;
      int n;
!     gss_channel_bindings_t bindings;
      struct gss_data *d = app_data;
  
      const char *knames[] = { "ftp", "host", NULL }, **kname = knames;
--- 242,248 ----
      int context_established = 0;
      char *p;
      int n;
!     gss_channel_bindings_t bindings=GSS_C_NO_CHANNEL_BINDINGS;
      struct gss_data *d = app_data;
  
      const char *knames[] = { "ftp", "host", NULL }, **kname = knames;
***************
*** 253,269 ****
      input.length = 0;
      input.value = NULL;
  
-     bindings = malloc(sizeof(*bindings));
  
!     sockaddr_to_gss_address (myctladdr,
  			     &bindings->initiator_addrtype,
  			     &bindings->initiator_address);
!     sockaddr_to_gss_address (hisctladdr,
  			     &bindings->acceptor_addrtype,
  			     &bindings->acceptor_address);
  
!     bindings->application_data.length = 0;
!     bindings->application_data.value = NULL;
  
      while(!context_established) {
         maj_stat = gss_init_sec_context(&min_stat,
--- 254,273 ----
      input.length = 0;
      input.value = NULL;
  
  
!     if (! gvFirewallNoChannelBinding) {
!        bindings = malloc(sizeof(*bindings));
! 
!        sockaddr_to_gss_address (myctladdr,
  			     &bindings->initiator_addrtype,
  			     &bindings->initiator_address);
!        sockaddr_to_gss_address (hisctladdr,
  			     &bindings->acceptor_addrtype,
  			     &bindings->acceptor_address);
  
!        bindings->application_data.length = 0;
!        bindings->application_data.value = NULL;
!     }
  
      while(!context_established) {
         maj_stat = gss_init_sec_context(&min_stat,
diff -rc yafc-1.0/src/ftp/ftpsend.c yafc-1.0-mm/src/ftp/ftpsend.c
*** yafc-1.0/src/ftp/ftpsend.c	2003-04-17 12:52:13.000000000 +0100
--- yafc-1.0-mm/src/ftp/ftpsend.c	2003-06-03 11:50:29.000000000 +0100
***************
*** 27,32 ****
--- 27,36 ----
  	char *e;
  	int i;
  
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         int sc; 
+ #endif
+ 
  	if(!ftp->has_pasv_command) {
  		ftp_err(_("Host doesn't support passive mode\n"));
  		return -1;
***************
*** 34,40 ****
--- 38,56 ----
  	ftp_set_tmp_verbosity(vbNone);
  
  	/* request passive mode */
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         if (gvFirewallClearCommandChannel) {
+            sc=ftp->sec_complete;
+            ftp->sec_complete = 0;
+         }
+ #endif
  	ftp_cmd("PASV");
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         if (gvFirewallClearCommandChannel) {
+             ftp->sec_complete = sc;
+             sc=0;
+         }
+ #endif
  
  	if(!ftp_connected())
  		return -1;
***************
*** 74,79 ****
--- 90,98 ----
  	struct sockaddr_in sa;
  	unsigned char *a, *p;
  	unsigned char pac[6];
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         int sc; 
+ #endif
  
  	if(!ftp_connected())
  		return -1;
***************
*** 97,104 ****
--- 116,135 ----
  		p = (unsigned char *)&ftp->data->local_addr.sin_port;
  
  		ftp_set_tmp_verbosity(vbError);
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+                 if (gvFirewallClearCommandChannel) {
+                    sc=ftp->sec_complete;
+                    ftp->sec_complete = 0;
+                 }
+ #endif
  		ftp_cmd("PORT %d,%d,%d,%d,%d,%d",
  				a[0], a[1], a[2], a[3], p[0], p[1]);
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+                 if (gvFirewallClearCommandChannel) {
+         	   ftp->sec_complete = sc;
+ 	            sc=0;
+                 }
+ #endif
  		if(ftp->code != ctComplete)
  			return -1;
  	}
***************
*** 778,783 ****
--- 809,817 ----
  	unsigned int old_reply_timeout;
  	Ftp *thisftp;
  	unsigned char addr[6];
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         int sc; 
+ #endif
  
  /*	printf("FxP: %s -> %s\n", srcftp->url->hostname, destftp->url->hostname);*/
  
***************
*** 805,812 ****
--- 839,858 ----
  	/* setup destination side */
  	ftp_use(destftp);
  	ftp_type(mode);
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         if (gvFirewallClearCommandChannel) {
+            sc=ftp->sec_complete;
+            ftp->sec_complete = 0;
+         }
+ #endif
  	ftp_cmd("PORT %d,%d,%d,%d,%d,%d",
  			addr[0], addr[1], addr[2], addr[3], addr[4], addr[5]);
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+         if (gvFirewallClearCommandChannel) {
+            ftp->sec_complete = sc;
+ 	   sc=0;
+         }
+ #endif
  	if(ftp->code != ctComplete) {
  		ftp_use(thisftp);
  		return -1;
diff -rc yafc-1.0/src/gvars.c yafc-1.0-mm/src/gvars.c
*** yafc-1.0/src/gvars.c	2003-04-17 12:52:14.000000000 +0100
--- yafc-1.0-mm/src/gvars.c	2003-10-20 18:02:49.000000000 +0100
***************
*** 169,174 ****
--- 169,179 ----
  #endif
  bool gvJmpBufSet = false;
  
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+ bool gvFirewallClearCommandChannel = false;
+ bool gvFirewallNoChannelBinding = false;
+ #endif
+ 
  void gvars_destroy(void)
  {
  	xfree(gvEditor);
diff -rc yafc-1.0/src/gvars.h yafc-1.0-mm/src/gvars.h
*** yafc-1.0/src/gvars.h	2003-04-17 12:52:14.000000000 +0100
--- yafc-1.0-mm/src/gvars.h	2003-10-20 18:02:49.000000000 +0100
***************
*** 165,170 ****
  #endif
  extern bool gvJmpBufSet;
  
! void gvars_destroy(void);
  
  #endif
--- 165,174 ----
  #endif
  extern bool gvJmpBufSet;
  
! #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
! extern bool gvFirewallClearCommandChannel;
! extern bool gvFirewallNoChannelBinding;
! #endif
  
+ void gvars_destroy(void);
  #endif
diff -rc yafc-1.0/src/rc.c yafc-1.0-mm/src/rc.c
*** yafc-1.0/src/rc.c	2003-04-17 12:52:15.000000000 +0100
--- yafc-1.0-mm/src/rc.c	2003-10-20 18:02:18.000000000 +0100
***************
*** 530,535 ****
--- 530,541 ----
  			parse_host(TRIG_DEFAULT, fp);
  		else if(strcasecmp(e, "local") == 0)
  			parse_host(TRIG_LOCAL, fp);
+ #if defined(HAVE_KRB4) || defined(HAVE_KRB5)
+ 		else if(strcasecmp(e, "FirewallClearCommandChannel") == 0)
+                         gvFirewallClearCommandChannel = nextbool(fp);
+ 		else if(strcasecmp(e, "FirewallNoChannelBinding") == 0)
+                         gvFirewallNoChannelBinding = nextbool(fp);
+ #endif
  		else
  			errp(_("Config parse error: '%s'\n"), e);
  	}
