diff -r -c heimdal-0.8/appl/ftp/ftp/cmds.c heimdal-0.8_mm/appl/ftp/ftp/cmds.c
*** heimdal-0.8/appl/ftp/ftp/cmds.c	2005-07-19 19:19:33.000000000 +0100
--- heimdal-0.8_mm/appl/ftp/ftp/cmds.c	2007-04-22 14:44:11.000000000 +0100
***************
*** 966,971 ****
--- 966,989 ----
  }
  
  /*
+  * Toggle fwccc mode on server.
+  */
+ /*VARARGS*/
+ void
+ setfwccc(int argc, char **argv)
+ {
+         if (command ("FWCCC") == COMPLETE) {
+           allowfwccc = !allowfwccc;
+           printf("FWCCC mode %s.\n", onoff(allowfwccc));
+           code = allowfwccc;
+         } else {
+           allowfwccc = 0;
+           printf("We don't support FWCCC mode, sorry.\n");
+           code = -1;
+       }
+ }
+ 
+ /*
   * Turn on interactive prompting
   * during mget, mput, and mdelete.
   */
diff -r -c heimdal-0.8/appl/ftp/ftp/cmdtab.c heimdal-0.8_mm/appl/ftp/ftp/cmdtab.c
*** heimdal-0.8/appl/ftp/ftp/cmdtab.c	2005-10-12 10:45:35.000000000 +0100
--- heimdal-0.8_mm/appl/ftp/ftp/cmdtab.c	2007-04-22 14:06:23.000000000 +0100
***************
*** 106,111 ****
--- 106,112 ----
  
  char	prothelp[] = 	"set protection level";
  char	prothelp_c[] =	"set command protection level";
+ char	fwccchelp[] =   "toggle fwccc mode";
  #ifdef KRB4
  char	kauthhelp[] = 	"get remote tokens";
  #endif
***************
*** 197,202 ****
--- 198,204 ----
  	{ "protect", 	prothelp, 	0, 	1, 	0,	sec_prot },
  	/* what MIT uses */
  	{ "cprotect",	prothelp_c,	0,	1,	1,	sec_prot_command },
+ 	{ "fwccc",      fwccchelp,      0,      0,      0,      setfwccc },
  #ifdef KRB4
  	{ "kauth", 	kauthhelp, 	0, 	1, 	0,	kauth },
  #endif
diff -r -c heimdal-0.8/appl/ftp/ftp/extern.h heimdal-0.8_mm/appl/ftp/ftp/extern.h
*** heimdal-0.8/appl/ftp/ftp/extern.h	2000-09-19 14:15:12.000000000 +0100
--- heimdal-0.8_mm/appl/ftp/ftp/extern.h	2007-04-22 14:07:18.000000000 +0100
***************
*** 150,155 ****
--- 150,156 ----
  void	syst (int, char **);
  void    tvsub (struct timeval *, struct timeval *, struct timeval *);
  void	user (int, char **);
+ void	setfwccc (int, char **);
  
  extern jmp_buf	abortprox;
  extern int	abrtflag;
diff -r -c heimdal-0.8/appl/ftp/ftp/ftp.c heimdal-0.8_mm/appl/ftp/ftp/ftp.c
*** heimdal-0.8/appl/ftp/ftp/ftp.c	2006-01-24 08:16:08.000000000 +0000
--- heimdal-0.8_mm/appl/ftp/ftp/ftp.c	2007-04-22 14:14:38.000000000 +0100
***************
*** 1208,1213 ****
--- 1208,1214 ----
  passive_mode (void)
  {
      int port;
+     int sc;
  
      data = socket (myctladdr->sa_family, SOCK_STREAM, 0);
      if (data < 0) {
***************
*** 1216,1227 ****
--- 1217,1233 ----
      }
      if (options & SO_DEBUG)
  	socket_set_debug (data);
+     sc=sec_complete;
+     if (allowfwccc)
+       sec_complete=0;
      if (command ("EPSV") != COMPLETE) {
  	if (command ("PASV") != COMPLETE) {
  	    printf ("Passive mode refused.\n");
+             sec_complete=sc;
  	    goto bad;
  	}
      }
+     sec_complete=sc;
  
      /*
       * Parse the reply to EPSV or PASV
***************
*** 1260,1265 ****
--- 1266,1272 ----
      int tmpno = 0;
      socklen_t len;
      int result;
+     int sc;
  
  noport:
      data_addr->sa_family = myctladdr->sa_family;
***************
*** 1316,1324 ****
--- 1323,1335 ----
  	if (debug == 0)
  	    verbose  = -1;
  
+         sc=sec_complete;
+         if (allowfwccc)
+             sec_complete=0;
  	result = command ("EPRT |%d|%s|%d|",
  			  inet_af, addr_str, 
  			  ntohs(socket_get_port (data_addr)));
+ 	sec_complete=sc;
  	verbose = overbose;
  
  	if (result == ERROR) {
***************
*** 1332,1337 ****
--- 1343,1351 ----
  		goto bad;
  	    }
  
+             sc=sec_complete;
+             if (allowfwccc)
+                 sec_complete=0;
  	    result = command("PORT %d,%d,%d,%d,%d,%d", 
  			     (a >> 24) & 0xff,
  			     (a >> 16) & 0xff,
***************
*** 1339,1344 ****
--- 1353,1359 ----
  			     a & 0xff,
  			     (p >> 8) & 0xff,
  			     p & 0xff);
+ 	    sec_complete=sc;
  	    if (result == ERROR && sendport == -1) {
  		sendport = 0;
  		tmpno = 1;
***************
*** 1552,1557 ****
--- 1567,1573 ----
      int secndflag = 0, prox_type, nfnd;
      char *cmd2;
      fd_set mask;
+     int sc;
  
      if (strcmp (cmd, "RETR"))
  	cmd2 = "RETR";
***************
*** 1565,1574 ****
--- 1581,1595 ----
      }
      if (curtype != prox_type)
  	changetype (prox_type, 1);
+     sc=sec_complete;
+     if (allowfwccc)
+       sec_complete=0;
      if (command ("PASV") != COMPLETE) {
  	printf ("proxy server does not support third party transfers.\n");
+ 	sec_complete=sc;
  	return;
      }
+     sec_complete=sc;
      pswitch (0);
      if (!connected) {
  	printf ("No primary connection\n");
***************
*** 1578,1587 ****
--- 1599,1613 ----
      }
      if (curtype != prox_type)
  	changetype (prox_type, 1);
+     sc=sec_complete;
+     if (allowfwccc)
+       sec_complete=0;
      if (command ("PORT %s", pasv) != COMPLETE) {
  	pswitch (1);
+ 	sec_complete=sc;
  	return;
      }
+     sec_complete=sc;
      if (setjmp (ptabort))
  	goto abort;
      oldintr = signal (SIGINT, abortpt);
diff -r -c heimdal-0.8/appl/ftp/ftp/ftp_var.h heimdal-0.8_mm/appl/ftp/ftp/ftp_var.h
*** heimdal-0.8/appl/ftp/ftp/ftp_var.h	2005-10-12 10:42:47.000000000 +0100
--- heimdal-0.8_mm/appl/ftp/ftp/ftp_var.h	2007-04-22 14:15:19.000000000 +0100
***************
*** 104,109 ****
--- 104,110 ----
  
  extern int	options;		/* used during socket creation */
  extern int      use_kerberos;           /* use Kerberos authentication */
+ extern int      allowfwccc;             /* send fwccc command before PORT/PASV/EPRT/EPSV command */
  
  /*
   * Format of command table.
diff -r -c heimdal-0.8/appl/ftp/ftp/globals.c heimdal-0.8_mm/appl/ftp/ftp/globals.c
*** heimdal-0.8/appl/ftp/ftp/globals.c	2005-10-12 10:42:32.000000000 +0100
--- heimdal-0.8_mm/appl/ftp/ftp/globals.c	2007-04-22 14:15:54.000000000 +0100
***************
*** 63,68 ****
--- 63,69 ----
  
  int	options;		/* used during socket creation */
  int     use_kerberos;           /* use Kerberos authentication */
+ int     allowfwccc;             /* send PORT/PASV and EPRT/EPSV commands in clear*/
  
  /*
   * Format of command table.
diff -r -c heimdal-0.8/appl/ftp/ftp/main.c heimdal-0.8_mm/appl/ftp/ftp/main.c
*** heimdal-0.8/appl/ftp/ftp/main.c	2005-10-12 10:42:20.000000000 +0100
--- heimdal-0.8_mm/appl/ftp/ftp/main.c	2007-04-22 14:16:40.000000000 +0100
***************
*** 72,77 ****
--- 72,79 ----
      { "encrypt", 'x', arg_flag, &doencrypt,
        "Encrypt command and data channel if possible" },
      { "version", 0,  arg_flag, &version_flag },
+     { "fwccc",   'c', arg_flag, &allowfwccc,
+       "Allow PORT/PASV/EPRT/EPASV commands in clear for stateful FW support with GSSAPI encryption", NULL},
      { "help",	'h', arg_flag, &help_flag },
  };
  
